File: scripts/dashboard_web/api_agents.py
"""API handler for /api/agents — real-time agent activity for brain visualization."""

import sys
from pathlib import Path

if __package__ in (None, ""):
    sys.path.insert(0, str(Path(__file__).resolve().parents[2]))
    from scripts.dashboard_web.middleware import query
else:
    from .middleware import query


def _build_sessions():
    """Build hierarchical session → children structure via single JOIN query."""
    rows = query(
        "SELECT s.agent_id AS session_id, s.agent_type AS session_type,"
        "  s.metadata AS session_meta,"
        "  c.agent_id AS child_id, c.agent_type AS child_type,"
        "  c.status AS child_status, c.description AS child_desc,"
        "  c.duration_s AS child_dur, c.model AS child_model,"
        "  c.started_at AS child_started"
        " FROM agent_activity s"
        " LEFT JOIN agent_activity c ON c.parent_session = s.agent_id"
        " WHERE s.agent_id LIKE 'session-%' AND s.status = 'running'"
        " ORDER BY s.agent_id, c.started_at DESC"
    )
    sess_map = {}
    for r in rows:
        sid = r["session_id"]
        if sid not in sess_map:
            parts = sid.split("-")  # session-{tool}-cli-{pid}
            pid = int(parts[-1]) if parts[-1].isdigit() else 0
            sess_map[sid] = {
                "session_id": sid, "type": r["session_type"],
                "pid": pid, "children": [],
            }
        if r.get("child_id"):
            sess_map[sid]["children"].append({
                "agent_id": r["child_id"], "type": r["child_type"],
                "status": r["child_status"], "description": r["child_desc"],
                "duration_s": r["child_dur"], "model": r["child_model"],
            })
    orphans = query(
        "SELECT agent_id, agent_type AS type, status, description,"
        "  duration_s, model"
        " FROM agent_activity"
        " WHERE parent_session IS NOT NULL"
        "  AND parent_session NOT IN"
        "    (SELECT agent_id FROM agent_activity WHERE agent_id LIKE 'session-%')"
        "  AND status = 'running'"
    )
    return list(sess_map.values()), orphans


def api_agents():
    """Return running agents, recent completions, sessions, and stats."""
    running = query(
        "SELECT agent_id, agent_type AS type, model, description,"
        " task_db_id, plan_id, host, region, parent_session,"
        " ROUND((julianday('now') - julianday(started_at)) * 86400, 1) AS duration_s"
        " FROM agent_activity WHERE status = 'running'"
        " ORDER BY started_at DESC"
    )
    recent = query(
        "SELECT agent_id, status, duration_s, tokens_total, cost_usd,"
        " agent_type AS type, model, completed_at, parent_session"
        " FROM agent_activity"
        " WHERE status IN ('completed', 'failed')"
        " AND completed_at >= datetime('now', '-1 hour')"
        " ORDER BY completed_at DESC LIMIT 20"
    )
    stats_row = query(
        "SELECT COALESCE(SUM(tokens_total), 0) AS total_tokens,"
        " COALESCE(SUM(cost_usd), 0) AS total_cost"
        " FROM agent_activity"
    )
    stats = stats_row[0] if stats_row else {"total_tokens": 0, "total_cost": 0}
    stats["active_count"] = len(running)
    ct = query(
        "SELECT COUNT(*) AS c FROM agent_activity"
        " WHERE status IN ('completed', 'failed')"
        " AND date(completed_at) = date('now')"
    )
    stats["completed_today"] = ct[0]["c"] if ct else 0
    by_model = query(
        "SELECT COALESCE(model, 'unknown') AS model,"
        " SUM(tokens_total) AS tokens"
        " FROM agent_activity GROUP BY model ORDER BY tokens DESC"
    )
    stats["by_model"] = {r["model"]: r["tokens"] for r in by_model}

    sessions, orphan_agents = _build_sessions()
    return {
        "running": running, "recent": recent, "stats": stats,
        "sessions": sessions, "orphan_agents": orphan_agents,
    }

File: rust/claude-core/src/server/api_agents.rs
use super::state::{query_rows, ApiError, ServerState};
use axum::extract::State;
use axum::routing::get;
use axum::{Json, Router};
use serde_json::{json, Value};

pub fn router() -> Router<ServerState> {
    Router::new()
        .route("/api/agents", get(api_agents))
        .route("/api/sessions", get(api_sessions))
        .route("/api/brain", get(api_brain))
}

async fn api_agents(State(state): State<ServerState>) -> Result<Json<Value>, ApiError> {
    let db = state.open_db()?;
    let running = query_rows(
        db.connection(),
        "SELECT agent_id, agent_type AS type, model, description, task_db_id, plan_id, host, region, parent_session, tokens_in, tokens_out, tokens_total, cost_usd, started_at, ROUND((julianday('now') - julianday(started_at)) * 86400, 1) AS duration_s FROM agent_activity WHERE status='running' ORDER BY started_at DESC",
        [],
    )
    .unwrap_or_default();
    let recent = query_rows(
        db.connection(),
        "SELECT agent_id, status, duration_s, tokens_total, cost_usd, agent_type AS type, model, completed_at, parent_session, description FROM agent_activity WHERE status IN ('completed','failed') AND completed_at >= datetime('now', '-1 hour') ORDER BY completed_at DESC LIMIT 20",
        [],
    )
    .unwrap_or_default();
    let by_model = query_rows(
        db.connection(),
        "SELECT COALESCE(model,'unknown') AS model, COALESCE(SUM(tokens_total),0) AS tokens FROM agent_activity GROUP BY model ORDER BY tokens DESC",
        [],
    )
    .unwrap_or_default();
    Ok(Json(json!({
        "running": running,
        "recent": recent,
        "stats": {"active_count": running.len(), "by_model": by_model}
    })))
}

async fn api_sessions(State(state): State<ServerState>) -> Result<Json<Value>, ApiError> {
    let db = state.open_db()?;
    let sessions = query_rows(
        db.connection(),
        "SELECT agent_id, agent_type AS type, description, status, metadata, started_at, tokens_total, cost_usd, model, ROUND((julianday('now') - julianday(started_at)) * 86400, 1) AS duration_s FROM agent_activity WHERE agent_id LIKE 'session-%' AND status='running' ORDER BY started_at",
        [],
    )
    .unwrap_or_default();
    Ok(Json(Value::Array(sessions)))
}

/// Consolidated endpoint for the neural graph: sessions + sub-agents + plans + tasks
async fn api_brain(State(state): State<ServerState>) -> Result<Json<Value>, ApiError> {
    let db = state.open_db()?;

    // Running sessions with full metadata
    let sessions = query_rows(
        db.connection(),
        "SELECT agent_id, agent_type AS type, description, status, metadata, started_at, tokens_in, tokens_out, tokens_total, cost_usd, model, ROUND((julianday('now') - julianday(started_at)) * 86400, 1) AS duration_s FROM agent_activity WHERE agent_id LIKE 'session-%' AND status='running' ORDER BY started_at",
        [],
    ).unwrap_or_default();

    // Running sub-agents with parent linkage
    let agents = query_rows(
        db.connection(),
        "SELECT agent_id, agent_type AS type, model, description, parent_session, task_db_id, plan_id, host, tokens_in, tokens_out, tokens_total, cost_usd, started_at, ROUND((julianday('now') - julianday(started_at)) * 86400, 1) AS duration_s FROM agent_activity WHERE status='running' AND agent_id NOT LIKE 'session-%' ORDER BY started_at DESC",
        [],
    ).unwrap_or_default();

    // Recent completed agents (last hour)
    let recent = query_rows(
        db.connection(),
        "SELECT agent_id, agent_type AS type, model, description, parent_session, status, tokens_total, cost_usd, duration_s, completed_at FROM agent_activity WHERE status IN ('completed','failed') AND completed_at >= datetime('now', '-1 hour') ORDER BY completed_at DESC LIMIT 30",
        [],
    ).unwrap_or_default();

    // Active plans with progress
    let plans = query_rows(
        db.connection(),
        "SELECT id, name, status, execution_host, tasks_done, tasks_total, ROUND(CASE WHEN tasks_total>0 THEN 100.0*tasks_done/tasks_total ELSE 0 END, 1) AS progress_pct FROM plans WHERE status='doing'",
        [],
    ).unwrap_or_default();

    // Tasks from active plans (all statuses for brain viz)
    let tasks = query_rows(
        db.connection(),
        "SELECT t.id, t.title, t.status, t.task_id, t.assignee, t.priority, t.type AS task_type, t.tokens, t.started_at, t.executor_session_id, t.executor_host, t.model, t.wave_id, t.output_data, p.name AS plan_name, p.id AS plan_id, w.name AS wave_name FROM tasks t JOIN plans p ON t.plan_id = p.id LEFT JOIN waves w ON t.wave_id_fk = w.id WHERE p.status = 'doing' ORDER BY t.status DESC, t.priority, t.id LIMIT 200",
        [],
    ).unwrap_or_default();

    // File changes from plan_commits (if any)
    let commits = query_rows(
        db.connection(),
        "SELECT pc.commit_sha, pc.commit_message, pc.lines_added, pc.lines_removed, pc.files_changed, pc.plan_id, pc.authored_at FROM plan_commits pc JOIN plans p ON pc.plan_id = p.id WHERE p.status = 'doing' OR pc.authored_at >= datetime('now', '-2 hours') ORDER BY pc.authored_at DESC LIMIT 20",
        [],
    ).unwrap_or_default();

    // Token totals per model (today)
    let token_summary = query_rows(
        db.connection(),
        "SELECT COALESCE(model,'unknown') AS model, COUNT(*) AS count, COALESCE(SUM(tokens_total),0) AS tokens, COALESCE(SUM(cost_usd),0) AS cost FROM agent_activity WHERE started_at >= date('now') GROUP BY model ORDER BY tokens DESC",
        [],
    ).unwrap_or_default();

    Ok(Json(json!({
        "sessions": sessions,
        "agents": agents,
        "recent": recent,
        "plans": plans,
        "tasks": tasks,
        "commits": commits,
        "token_summary": token_summary
    })))
}
